backend_py/primary/primary/services/sumo_access/vfp_types.py

```diff
@@ -0,0 +1,348 @@
+    WTF = "WTF"
+
+
+# Gas fraction types
+class GFR(Enum):
+    GOR = "GOR"
+    GLR = "GLR"
+    OGR = "OGR"
+    MMW = "MMW"
+
+
+# Artificial lift types
+class ALQ(Enum):
+    GRAT = "GRAT"
+    IGLR = "IGLR"
+    TGLR = "TGLR"
+    PUMP = "PUMP"
+    COMP = "COMP"
+    DENO = "DENO"
+    DENG = "DENG"
+    BEAN = "BEAN"
+    UNDEFINED = "''"
+
+
+# Unit types
+class UnitType(Enum):
+    METRIC = "METRIC"
+    FIELD = "FIELD"
+    LAB = "LAB"
+    PVTM = "PVT-M"
+    DEFAULT = "DEFAULT"
+
+
+# Tabulated values type
+class TabType(Enum):
+    BHP = "BHP"
+    THT = "TEMP"
+
+
+# The length of bhp_values is len(thp_values)*len(wfr_values)*len(gfr_values)*len(alq_values)*len(flow_rate_values)
+# The values are ordered so that the index of flow_rate_values moves fastest, and the index of thp_values moves
+# slowest. The order is: THP, WFR, GFR, ALQ, Flow rates.
+class VfpProdTable(BaseModel):
+    vfp_type: VfpType
+    table_number: int
+    datum: float
+    thp_type: THP
+    wfr_type: WFR
+    gfr_type: GFR
+    alq_type: ALQ
+    flow_rate_type: FlowRateTypeProd
+    unit_type: UnitType
+    tab_type: TabType
+    thp_values: List[float]
+    wfr_values: List[float]
+    gfr_values: List[float]
+    alq_values: List[float]
+    flow_rate_values: List[float]
+    bhp_values: List[float]
+    flow_rate_unit: str
+    thp_unit: str
+    wfr_unit: str
+    gfr_unit: str
+    alq_unit: str
+    bhp_unit: str
+
+
+class VfpInjTable(BaseModel):
+    vfp_type: VfpType
+    table_number: int
+    datum: float
+    flow_rate_type: FlowRateTypeInj
+    unit_type: UnitType
+    tab_type: TabType
+    thp_values: List[float]
+    flow_rate_values: List[float]
+    bhp_values: List[float]
+    flow_rate_unit: str
+    thp_unit: str
+    bhp_unit: str
+
+
+# Unit definitions for VFPPROD
+VFPPROD_UNITS: Dict[UnitType, Dict[VfpParam, Any]] = {
+    UnitType.DEFAULT: {
+        VfpParam.FLOWRATE: {
+            FlowRateTypeProd.OIL: "",
+            FlowRateTypeProd.LIQ: "",
+            FlowRateTypeProd.GAS: "",
+            FlowRateTypeProd.WG: "",
+            FlowRateTypeProd.TM: "",
+        },
+        VfpParam.THP: {THP.THP: "barsa"},
+        VfpParam.WFR: {
+            WFR.WOR: "",
+            WFR.WCT: "",
+            WFR.WGR: "",
+            WFR.WWR: "",
+            WFR.WTF: "",
+        },
+        VfpParam.GFR: {
+            GFR.GOR: "",
+            GFR.GLR: "",
+            GFR.OGR: "",
+            GFR.MMW: "",
+        },
+        VfpParam.ALQ: {
+            ALQ.GRAT: "",
+            ALQ.IGLR: "",
+            ALQ.TGLR: "",
+            ALQ.DENO: "",
+            ALQ.DENG: "",
+            ALQ.BEAN: "",
+            ALQ.UNDEFINED: "",
+        },
+    },
+    UnitType.METRIC: {
+        VfpParam.FLOWRATE: {
+            FlowRateTypeProd.OIL: "sm3/day",
+            FlowRateTypeProd.LIQ: "sm3/day",
+            FlowRateTypeProd.GAS: "sm3/day",
+            FlowRateTypeProd.WG: "sm3/day",
+            FlowRateTypeProd.TM: "kg-M/day",
```

Comment:
Out of interest, what does the `M` unit indicate?

Reply:
I had to look it up, it seems to be molecular weight. I have never seen it used :)